model/extensions/Stateen/stateen_regs.sail

```diff
@@ -0,0 +1,396 @@
+  if not(enable_mstateen()) then ones()
+  else match idx {
+    0 => mstateen0.bits,
+    1 => mstateen1.bits,
+    2 => mstateen2.bits,
+    3 => mstateen3.bits,
+  }
+}
+
+mapping clause csr_name_map = 0x30C <-> "mstateen0"
+mapping clause csr_name_map = 0x30D <-> "mstateen1"
+mapping clause csr_name_map = 0x30E <-> "mstateen2"
+mapping clause csr_name_map = 0x30F <-> "mstateen3"
+
+mapping clause csr_name_map = 0x31C <-> "mstateen0h"
+mapping clause csr_name_map = 0x31D <-> "mstateen1h"
+mapping clause csr_name_map = 0x31E <-> "mstateen2h"
+mapping clause csr_name_map = 0x31F <-> "mstateen3h"
+
+function clause is_CSR_accessible(0x30C, _, _) = enable_mstateen()
+function clause is_CSR_accessible(0x30D, _, _) = enable_mstateen()
+function clause is_CSR_accessible(0x30E, _, _) = enable_mstateen()
+function clause is_CSR_accessible(0x30F, _, _) = enable_mstateen()
+
+function clause is_CSR_accessible(0x31C, _, _) = enable_mstateen() & (xlen == 32)
+function clause is_CSR_accessible(0x31D, _, _) = enable_mstateen() & (xlen == 32)
+function clause is_CSR_accessible(0x31E, _, _) = enable_mstateen() & (xlen == 32)
+function clause is_CSR_accessible(0x31F, _, _) = enable_mstateen() & (xlen == 32)
+
+function clause read_CSR(0x30C) = mstateen0.bits[xlen - 1 .. 0]
+function clause read_CSR(0x30D) = mstateen1.bits[xlen - 1 .. 0]
+function clause read_CSR(0x30E) = mstateen2.bits[xlen - 1 .. 0]
+function clause read_CSR(0x30F) = mstateen3.bits[xlen - 1 .. 0]
+
+function clause read_CSR(0x31C if xlen == 32) = mstateen0.bits[63 .. 32]
+function clause read_CSR(0x31D if xlen == 32) = mstateen1.bits[63 .. 32]
+function clause read_CSR(0x31E if xlen == 32) = mstateen2.bits[63 .. 32]
+function clause read_CSR(0x31F if xlen == 32) = mstateen3.bits[63 .. 32]
+
+function clause write_CSR((0x30C, value) if xlen == 32) = { mstateen0 = legalize_mstateen0(mstateen0, mstateen0.bits[63 .. 32] @ value); Ok(mstateen0.bits[31 .. 0]) }
+function clause write_CSR((0x30D, value) if xlen == 32) = { mstateen1 = legalize_mstateen1(mstateen1, mstateen1.bits[63 .. 32] @ value); Ok(mstateen1.bits[31 .. 0]) }
+function clause write_CSR((0x30E, value) if xlen == 32) = { mstateen2 = legalize_mstateen2(mstateen2, mstateen2.bits[63 .. 32] @ value); Ok(mstateen2.bits[31 .. 0]) }
+function clause write_CSR((0x30F, value) if xlen == 32) = { mstateen3 = legalize_mstateen3(mstateen3, mstateen3.bits[63 .. 32] @ value); Ok(mstateen3.bits[31 .. 0]) }
+
+function clause write_CSR((0x30C, value) if xlen == 64) = { mstateen0 = legalize_mstateen0(mstateen0, value); Ok(mstateen0.bits) }
+function clause write_CSR((0x30D, value) if xlen == 64) = { mstateen1 = legalize_mstateen1(mstateen1, value); Ok(mstateen1.bits) }
+function clause write_CSR((0x30E, value) if xlen == 64) = { mstateen2 = legalize_mstateen2(mstateen2, value); Ok(mstateen2.bits) }
+function clause write_CSR((0x30F, value) if xlen == 64) = { mstateen3 = legalize_mstateen3(mstateen3, value); Ok(mstateen3.bits) }
```

Comment:
Perhaps ignore writes to `xstateen{1-3}` for `x` in `h,s,m` since they don't have any useful bits and SE1-3 are initialized to zero anyway, and make them read-only zero?